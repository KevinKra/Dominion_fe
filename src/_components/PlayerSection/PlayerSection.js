import { connect } from "react-redux";
import React, { Component } from "react";
import * as actions from "../../_redux/actions";
import ActivatedCards from "../ActivatedCards/ActivatedCards";
import PlayerDeck from "../PlayerDeck/PlayerDeck";
import PlayerHand from "../PlayerHand/PlayerHand";
import DiscardPile from "../DiscardPile/DiscardPile";
import { updateGameState, updatePlayerState } from "../../_utils/apiCalls";
import "./PlayerSection.scss";

export class PlayerSection extends Component {
  state = {
    dataUpdated: false
  };

  componentDidMount = () => {
    this.turnInterval = setInterval(() => this.requestPlayerTurn(), 15000);
    this.requestPlayerTurn();
  };

  componentWillUnmount = () => {
    clearInterval(this.turnInterval);
  };

  requestPlayerTurn = async () => {
    const gameState = await updateGameState(this.props.gameID);
    if (gameState.tableDeck.length === 0) {
      return console.log("Waiting for game to start.");
    }
    if (gameState.activePlayerId === this.props.playerID) {
      //active player
      console.log(
        "It's your turn, turnInterval turned off. Please complete your turn."
      );
      clearInterval(this.turnInterval);
      this.updatePlayerData();
    } else {
      //waiting player
      console.log("Not your turn, continuing the interval check the turn");
      this.updatePlayerData();
    }
    console.log(gameState);
  };

  draw = deck => {
    const newHand = deck.splice(0, 5);
    return { newHand, deck };
  };

  updatePlayerData = async () => {
    if (!this.state.dataUpdated) {
      const playerData = await updatePlayerState(
        this.props.gameID,
        this.props.playerID
      );
      console.log(playerData);
      const drawnCards = this.draw(playerData.deck);
      this.props.updatePlayerCards(
        drawnCards.deck,
        drawnCards.newHand,
        playerData.discardPile || null
      );
    }
    this.setState({ dataUpdated: true });
  };

  //!!! need to reset dataUpdated to false at the start of every turn so players will fetch
  //their data once regardless whether they are active or waiting
  startNewTurn() {
    //this function needs to reset dataUpdatedToFalse
    //needs to be called at the start of requestPlayerTurn to reset local state
    //then both players are able to fetch their respective cards and toggle that state to off.
  }

  render() {
    return (
      <section className="PlayerSection">
        <ActivatedCards />
        <PlayerDeck />
        <PlayerHand />
        <DiscardPile />
      </section>
    );
  }
}

export const mapStateToProps = state => ({
  playerTurn: state.playerTurn,
  playerDeck: state.playerDeck,
  playerHand: state.playerHand,
  discardPile: state.discardPile,
  gameID: state.gameID,
  playerID: state.playerID
});

export const mapDispatchToProps = dispatch => ({
  startTurn: () => dispatch(actions.beginTurn()),
  cycleToPhase: phase => dispatch(actions.cycleToPhase(phase)),
  endTurn: () => dispatch(actions.endTurn()),
  updatePlayerCards: (deck, hand, discardPile) =>
    dispatch(actions.updatePlayerCards(deck, hand, discardPile))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PlayerSection);
