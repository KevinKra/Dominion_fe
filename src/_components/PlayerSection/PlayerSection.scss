@import "../../styles/_variables.scss";
@import "../../styles/_animations.scss";

.PlayerSection {
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: 100%;
  background-color: $color1;
  .end-turn {
    position: absolute;
    right: 1.5rem;
    bottom: 1.5rem;
    padding: 0.5rem;
    border: 1px solid $color2;
    background-color: $color2;
    color: $color1;
    cursor: pointer;
    transition: 0.49s cubic-bezier(0.4, 2.42, 0.29, -0.04);
    &:active {
      background-color: white;
      color: $color2;
    }
    &:focus {
      outline: none;
    }
  }
}

.loading-page {
  font-family: "Germania One", cursive;
  border: 1px solid blue;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
  .loading-container {
    animation: fadeInUp 1.2s;
    position: relative;
    z-index: 2;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    height: 225px;
    width: 350px;
    border: 1px solid $color2;
    border-top: 1px dashed $color2;
    border-bottom: 1px dashed $color2;
    background-color: $color1;
    box-shadow: 1px 1px 15px $color2;
  }
  .loading-icon-container {
    flex: 1;
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: $color2;
    width: 100%;
  }
  .loading-icon {
  }
  .game-id {
  }
  .background-image {
    position: absolute;
    z-index: 1;
    height: 100vh;
    width: 100vw;
    background: url("https://tsdbcg.herokuapp.com/card_images/freshwater-bay.jpg");
    background-size: cover;
  }
  h2 {
    margin: 1rem 0;
  }
}
