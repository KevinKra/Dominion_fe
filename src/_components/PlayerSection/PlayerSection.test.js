import React from "react";
import { PlayerSection } from "./PlayerSection";
import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";

describe("<PlayerSection />", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<PlayerSection playerTurn={true} tableCards={1} />);
  });
  it("should match the snapshot", () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
