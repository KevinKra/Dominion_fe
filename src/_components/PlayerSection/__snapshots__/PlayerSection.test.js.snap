// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayerSection /> should match the snapshot 1`] = `
<section
  className="PlayerSection"
>
  <ActivatedCards />
  <Connect(PlayerDeck) />
  <PlayerHand />
  <DiscardPile />
</section>
`;
